### PR TITLE
WIP: Add gentkn cli command

### PR DESCRIPTION
**Pull request by nstogner**
**WORK IN PROGRESS**

Add a cli tool for generating JWT's.

### CODE REVIEW

File: cmd/gentkn/main.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	var cfg struct {
+		KeyID string `envconfig:"KEY_ID" default:"default"`
+
+		Claims struct {
+			Audience string   `envconfig:"AUDIENCE"`
+			Subject  string   `envconfig:"SUBJECT"`
+			Roles    []string `envconfig:"ROLES"`
+		} `envconfig:"CLAIMS"`
+	}
+	if err := envconfig.Process("GENTKN", &cfg); err != nil {
```
Comment (ardan-bkennedy) [inline cmd/gentkn/main.go]: Why are we not using our internal flag package? Then you also get the command line options for free. If there are bugs let's fix them. I know I have one pending.

File: cmd/gentkn/main.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	if err := flag.Process(&cfg); err != nil {
+		if err != flag.ErrHelp {
+			log.Fatalf("main : Parsing Command Line : %v", err)
+		}
+		return // We displayed help.
+	}
+	*/
+
+	// Source the private key.
+	var keyRdr io.Reader
```
Comment (ardan-bkennedy) [inline cmd/gentkn/main.go]: This variable declaration should move down.

File: cmd/gentkn/main.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	if len(os.Args) < 2 {
+		log.Fatal("main : Missing Required Param <key-file>")
+	}
+	switch file := os.Args[1]; file {
```
Comment (ardan-bkennedy) [inline cmd/gentkn/main.go]: I think we need a line feed here and it's probably a good place for that variable.

File: cmd/gentkn/main.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	if err != nil {
+		log.Fatalf("main : Reading Key : %v", err)
+	}
+
+	key, err := jwt.ParseRSAPrivateKeyFromPEM(keyContents)
+	if err != nil {
+		log.Fatalf("main : Parsing RSA Private Key : %v", err)
+	}
+
+	// Build the claims.
```
Comment (ardan-bkennedy) [inline cmd/gentkn/main.go]: Not sure this comment is adding value.

File: cmd/gentkn/main.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+			IssuedAt:  now.Unix(),
+			Subject:   cfg.Claims.Subject,
+		},
+		Roles: cfg.Claims.Roles,
+	}
+	if err := claims.Valid(); err != nil {
+		log.Fatalf("main : Invalid Claims : %v", err)
+	}
+
+	// Generate and sign the token.
```
Comment (ardan-bkennedy) [inline cmd/gentkn/main.go]: Not sure the comment is adding value.

File: cmd/gentkn/main.go (hunk 1 above)
Comment (ardan-bkennedy) [reply cmd/gentkn/main.go]: I get that, I want to know why we are passing, non-flagged arguments?

Comment (nstogner) [reply cmd/gentkn/main.go]: Seems to go against unix conventions to have required flags for cli tools

Comment (nstogner) [reply cmd/gentkn/main.go]: That coupled with not really having a good default for that value, seems proper to use a non-flag arg

**Comment by jcbwlkr** [comment]
Closing per conversation with @nstogner. The new PR that adds a token endpoint is preferred

**Pull request closed without merging**